File: Python/046__Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        def once(lists, n):
            result = []
            for lst in lists:
                for i in range(len(lst) + 1):
                    result.append(lst[:i] + [n] + lst[i:])
            return result

        result = [[nums[0]]]
        for i in range(1, len(nums)):
            result = once(result, nums[i])
        return result

File: Java/046__Permutations.java
class Solution {

    List<List<Integer>> result = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        List<Integer> nums_lst = new ArrayList<>();
        for (int val : nums)
            nums_lst.add(val);
        backtrack(nums.length, nums_lst, 0);
        return result;
    }

    public void backtrack(int n, List<Integer> nums, int index) {
        if (index == n)
            result.add(new ArrayList<Integer>(nums));

        for (int i = index; i < n; i++) {
            // place i-th integer first
            // in the current permutation
            Collections.swap(nums, index, i);
            // use next integers to complete the permutations
            backtrack(n, nums, index + 1);
            // backtrack
            Collections.swap(nums, index, i);
         }
    }
}
